### ISSUE TITLE
Variable x not accessed can be prevented with __x__ (and not _x)

### ISSUE THREAD

**Issue by danr** [microsoft/pyright#2461]
**Describe the bug**
Variable not accessed warnings can be prevented with double leading and trailing underscores.
Variable not accessed warnings cannot be prevented with a single leading underscore.

**To Reproduce**
```python
def example(s: str):
    t     = s # Variable t     not accessed
    _t    = s # Variable _t    not accessed (?!)
    __t   = s # Variable __t   not accessed
    t_    = s # Variable t_    not accessed
    _t_   = s # Variable _t_   not accessed
    __t_  = s # Variable __t_  not accessed
    t__   = s # Variable t__   not accessed
    _t__  = s # Variable _t__  not accessed
    __t__ = s # no warning (?!)
    return s
```
Steps to reproduce the behavior.

**Expected behavior**
Either:
1. Variable not accessed warnings _can_ be prevented with a single leading underscore.

Or:

2. Variable not accessed warnings _cannot_ be prevented with double leading and trailing underscores.

**Comment by erictraut** [comment]
Pyright doesn't generate warnings or errors for unreferenced variables. It generates hints that editors can either ignore or use to display the symbol in an appropriately stylized manner, like a grayed-out appearance. 

Dundered symbols are intentionally exempted from this check because dundered names often have implied uses. For example, `__all__` is typically not referenced explicitly. The behavior you're seeing is intended.

**Comment by danr** [comment]
Hi @erictraut, thanks for your reply. One thing caught my attention:

> Pyright doesn't generate warnings or errors for unreferenced variables. 

This does not seem correct. Consider running pyright like this:
```sh
#!/bin/sh
>example.py printf %s '
def example(s: str):
    t     = s # Variable t     not accessed
    _t    = s # Variable _t    not accessed (?!)
    __t   = s # Variable __t   not accessed
    t_    = s # Variable t_    not accessed
    _t_   = s # Variable _t_   not accessed
    __t_  = s # Variable __t_  not accessed
    t__   = s # Variable t__   not accessed
    _t__  = s # Variable _t__  not accessed
    __t__ = s # no warning (?!)
    return s
'
>pyrightconfig.json printf %s '{"typeCheckingMode":"strict"}'
pyright example.py
```
Running this does make pyright generate errors:
```
$ ./play.sh 
Loading configuration file at /tmp/pyrightconfig.json
Assuming Python version 3.9
Assuming Python platform Linux
Auto-excluding **/node_modules
Auto-excluding **/__pycache__
Auto-excluding .git
stubPath /tmp/typings is not a valid directory.
Searching for source files
Found 1 source file
/tmp/example.py
  /tmp/example.py:3:5 - error: Variable "t" is not accessed (reportUnusedVariable)
  /tmp/example.py:4:5 - error: Variable "_t" is not accessed (reportUnusedVariable)
  /tmp/example.py:5:5 - error: Variable "__t" is not accessed (reportUnusedVariable)
  /tmp/example.py:6:5 - error: Variable "t_" is not accessed (reportUnusedVariable)
  /tmp/example.py:7:5 - error: Variable "_t_" is not accessed (reportUnusedVariable)
  /tmp/example.py:8:5 - error: Variable "__t_" is not accessed (reportUnusedVariable)
  /tmp/example.py:9:5 - error: Variable "t__" is not accessed (reportUnusedVariable)
  /tmp/example.py:10:5 - error: Variable "_t__" is not accessed (reportUnusedVariable)
8 errors, 0 warnings, 0 infos 
Completed in 0.543sec
```
The pyright language server also reports these as diagnostics.

**Comment by erictraut** [comment]
Yes, if you explicitly enable `reportUnusedVariable` (or implicitly enable it, as implied by strict mode), then pyright will generate an error. By default it does not. I thought you were referring to the default behavior, but in rereading your initial post, you may have been referring to the behavior with `reportUnusedVariable` enabled.

The current rules used by `reportUnusedVariable` are as follows:
* It is never generated for the input parameter to a function or method
* It is never generated for a single underscore, since that is conventionally used in Python to indicate that a variable's value is meant to go unused
* It uses a dundered name, since those are used by the Python runtime & library implementations for various special cases

It sounds like you would like to apply different principles for this rule. Can you provide more details behind the principles you have suggested above? Are there any documented specifications or conventions that support these principles?

**Comment by danr** [comment]
Thank you for your thoughful response. Indeed I would like to apply different principles, but looking around at mandated Python resources (like PEP8), this principle does indeed not seem to "exist" in python. I understand that it's outside the scope of the pyright project to enforce my specific principles, so I rest my case now :)

**Comment by danr** [comment]
Seems like you changed this anyway! https://github.com/microsoft/pyright/commit/ced19cbcbd625b9fcc1a807c8239869a2badb43b

Thanks! :+1:
